packages/@ember/debug/lib/deprecate.ts

```diff
@@ -187,6 +192,28 @@ if (DEBUG) {
     assert(missingOptionsIdDeprecation, Boolean(options!.id));
     assert(missingOptionsUntilDeprecation, Boolean(options!.until));
 
+    if (!options.for) {
+      deprecate(missingOptionsForDeprecation, Boolean(options!.for), {
+        id: 'ember-source.deprecation-without-for',
+        until: '4.0.0',
+        for: 'ember-source',
+        since: {
+          available: '3.22.0',
+        },
+      });
+    }
```

Comment:
Based on our convo today, let's wrap these in another guard to ensure we only issue a deprecation once for missing `for` for a specific `id`. I'm thinking something like:

```js
    if (!options.for && !FOR_MISSING_DEPRECATIONS.has(options.id)) {
      FOR_MISSING_DEPRECATION.add(options.id);
      
      deprecate(missingOptionsForDeprecation, Boolean(options!.for), {
        id: 'ember-source.deprecation-without-for',
        until: '4.0.0',
        for: 'ember-source',
        since: {
          available: '3.22.0',
        },
      });
    }
```

This avoids issuing the `for` missing deprecation for **every** invocation (which could be a massive number).

Reply:
Also, we should ensure that we include information on the original `options.id` in the deprecation message. At the moment, I think we are using a static `missingOptionsForDeprecation` message, but I'd like to see the original deprecation `id` in the message.

Reply:
✅